src/storage/src/mem_table.rs

```diff
@@ -182,3 +221,395 @@ impl KeyOp {
+                false
+            } else {
+                // Key has exceeded the current key range. Advance to the next range.
+                loop {
+                    range = delete_ranges_iter.next();
+                    if let Some((range_start, range_end)) = range {
+                        assert!(
+                            range_start <= range_end,
+                            "range_end {:?} smaller than range_start {:?}",
+                            range_start,
+                            range_end
+                        );
+                        if key < range_start {
+                            // Not fall in the next delete range
+                            break true;
+                        } else if key < range_end {
+                            // Fall in the next delete range
+                            break false;
+                        } else {
+                            // Exceed the next delete range. Go to the next delete range if there is
+                            // any in the next loop
+                            continue;
+                        }
+                    } else {
+                        // No more delete range.
+                        break true;
+                    }
+                }
+            }
+        } else {
+            true
+        }
+    })
+}
+
+impl<S: StateStoreWrite + StateStoreRead> LocalStateStore for MemtableLocalStateStore<S> {
+    type FlushFuture<'a> = impl Future<Output = StorageResult<usize>> + 'a;
+    type GetFuture<'a> = impl GetFutureTrait<'a>;
+    type IterFuture<'a> = impl Future<Output = StorageResult<Self::IterStream<'a>>> + Send + 'a;
+    type IterStream<'a> = impl StateStoreIterItemStream + 'a;
+
+    fn get<'a>(&'a self, key: &'a [u8], read_options: ReadOptions) -> Self::GetFuture<'_> {
+        async move {
+            match self.mem_table.buffer.get(key) {
+                None => self.inner.get(key, self.epoch(), read_options).await,
+                Some(op) => match op {
+                    KeyOp::Insert(value) | KeyOp::Update((_, value)) => Ok(Some(value.clone())),
+                    KeyOp::Delete(_) => Ok(None),
+                },
+            }
+        }
+    }
+
+    fn iter(
+        &self,
+        key_range: (Bound<Vec<u8>>, Bound<Vec<u8>>),
+        read_options: ReadOptions,
+    ) -> Self::IterFuture<'_> {
+        async move {
+            let stream = self
+                .inner
+                .iter(key_range.clone(), self.epoch(), read_options)
+                .await?;
+            let (l, r) = key_range;
+            let key_range = (l.map(Bytes::from), r.map(Bytes::from));
+            Ok(merge_stream(
+                self.mem_table.iter(key_range),
+                stream,
+                self.table_id,
+                self.epoch(),
+            ))
+        }
+    }
+
+    fn insert(&mut self, key: Bytes, new_val: Bytes, old_val: Option<Bytes>) -> StorageResult<()> {
```

Comment:
Will `upsert` be a better naming?

Reply:
IIUC, now the `upsert` is the `insert` on `Memtable` with `is_consistent_op = false`?